### ISSUE TITLE
Indentation of visualforce tags doesn't work

### ISSUE THREAD

**Issue by kgrzywacz** [neowit/vim-force.com#67]
Hi @neowit,

Recently it struck me that indentation with `=` operator doesn't work for VF tags. I've looked trough `.indent/visualforce.vim` as well as `./runtime/indent/html.vim` and it seems that only tags hardcoded in `html.vim` are indented, and there is no nice way of reusing logic from this file, due to script scope of variable holding the tags. 

Even if there was a way of injecting tags to `s:indent_tags` of `html.vim` there are two problems:

1. Logic that is in place ignores `:` sign, and everything after, so all apex tags are read as `apex` instead of i.e. `apex:page`.
2. Most of VF tags can be self-closing, which is not really supported by logic in `html.vim` - either tag is indented or not(example `<p/>` tag won't cause indentation of following lines, but also everything between tags `<p>` and `</p>` won't be indented.

My questions are:

Am I missing something?
Did it ever worked?
Did you consider working on script for indentation of VF tags?

**Comment by neowit** [comment]
Hello @kgrzywacz

Thank you for the report.
I did not dig much into indentation for VF tags and do not think it has ever worked.
Would be happy to accept a pull request if someone comes up with a way to fix VF tags indentation.

FYI: vim-coffee-script has an example of wrapping the default html indentation: https://github.com/kchmck/vim-coffee-script/blob/master/after/indent/html.vim
perhaps something similar can be done for VF indentation.

**Comment by kgrzywacz** [comment]
Thanks @neowit, although example you provided combines two existing indentation scripts for one filetype. 

I will look into that, and maybe come up with something.

**Comment by neowit** [comment]
I should have been clearer - the reason I mentioned vim-coffee-script example is precisely because it combines two existing indentations.
`apex:something` tags are supported nicely by XML indentation while everything else can be handled by the HTML one.
Try setting `set ft=xml` in your VF buffer - you may observe that `apex:...` and all HTML tags get (more or less) properly aligned, but things like `<script>` & `<style>` are not.

**Comment by kgrzywacz** [comment]
Hi @neowit,
I was playing with indenting of VF, but something strange is happening. 
When I set filetype of VF file to xml or ant - it is correctly indented (at least \<tags\>) but when I change my [indent/visualforce.vim](https://github.com/neowit/vim-force.com/blob/master/indent/visualforce.vim) to:
```
if exists("b:did_indent")
  finish
endif

" Use XML formatting rules
runtime! indent/xml.vim
```
(This is basically content of [indent/ant.vim](https://github.com/vim/vim/blob/master/runtime/indent/ant.vim))
and indent following file:
```
<apex:page>
<apex:pageBlock>
<apex:pageBlockSection>
<apex:outputText>
</apex:outputText>
</apex:pageBlockSection>
</apex:pageBlock>
</apex:page>
```
Nothing changes.
If we try the same with following text(all lines have one tab at the beginning:
```
	<apex:page>
	<apex:pageBlock>
	<apex:pageBlockSection>
	<apex:outputText>
	</apex:outputText>
	</apex:pageBlockSection>
	</apex:pageBlock>
	</apex:page>
```
after `gg=G` we end up with correctly indented file:
```
<apex:page>
	<apex:pageBlock>
		<apex:pageBlockSection>
			<apex:outputText>
			</apex:outputText>
		</apex:pageBlockSection>
	</apex:pageBlock>
</apex:page>
```

Can't realy see why this happens, if I change filetype to xml or ant, it is always indented correctly, any idea?

**Comment by neowit** [comment]
At first glance it looks like some kind of quirk of `indent/xml.vim` because after "indentation" (which does nothing) it still reports `9 lines indented`.

The reason why indentation works with `set ft=xml` is probably related to the fact that `indent/xml.vim` is not the only (xml related) vim file loaded with `set ft=xml`.

Another interesting thing I noticed is that by default `set indentexpr?` returns 
> indentexpr=XmlIndentGet(v:lnum,1)

If I change it to ignore syntax
```
set indentexpr=XmlIndentGet(v:lnum,0)
```
then your first example (with no space/tab in front of each line) gets indented as expected.

**Comment by neowit** [comment]
Hello @kgrzywacz 

Thank you for the update. Looks good.

Out of interest - does it actually need `runtime indent/css.vim` after `runtime indent/html.vim`. Does not `indent/html.vim` include logic for `indent/css.vim` ?

**Comment by kgrzywacz** [comment]
Hi @neowit 

html.vim script for indention use neither css.vim nor javascript.vim, it implements it's own logic for _alien_ content. ([Javascript](https://github.com/vim/vim/blob/master/runtime/indent/html.vim#L586), [CSS](https://github.com/vim/vim/blob/master/runtime/indent/html.vim#L606))

I used css.vim instead of html.vim because there is a bug in in html.vim logic for css content, which results in adding additional tab for every new line(doesn't matter if you just `<c-r>` in insert mode or `o` in normal).

it will look like that:
```
.test {
	background-color: red;
}
```
```
.test {
	background-color: red;
		color: white;
}
```
```
.test {
	background-color: red;
		color: white;
			display:block;
}
```
And so on.

If you indent these lines with `=`, it will recalculate and behave as expected - each line inside test class from example will be indented with only one tab.

I prefered to use less buggy logic, therefore I used css.vim

I've also considered using javascript.vim or cindent() function for javascript(which is used in html.vim, with some additional logic), but both happen to not work correctly at all times.
I will try to use one of them in future, because loading whole html.vim only for javascript isn't ideal(but it works).
